x-pack/plugin/ccr/src/internalClusterTest/java/org/elasticsearch/xpack/ccr/CcrRepositoryIT.java

```diff
@@ -613,83 +602,76 @@ public void testCcrRepositoryFailsToFetchSnapshotShardSizes() throws Exception {
-            assertThat(repository.getMetadata().type(), equalTo(CcrRepository.TYPE));
-            assertThat(repository.getMetadata().name(), equalTo(leaderCluster));
-
-            for (int i = 0; i < numberOfShards; i++) {
-                final Index index = indexMetadata.getIndex();
-                final int shardId = i;
-                ElasticsearchException exception = expectThrows(ElasticsearchException.class,
-                    () -> repository.getShardSnapshotStatus(
-                        new SnapshotId(CcrRepository.LATEST, CcrRepository.LATEST),
-                        new IndexId(index.getName(), index.getUUID()),
-                        new ShardId(index, shardId)));
-                assertThat(exception.getMessage(), equalTo("simulated"));
-            }
-            assertThat(indicesStatsRequestsCount.getAndSet(0), equalTo(numberOfShards));
-
-            final RestoreService restoreService = getFollowerCluster().getCurrentMasterNodeInstance(RestoreService.class);
-            final ClusterService clusterService = getFollowerCluster().getCurrentMasterNodeInstance(ClusterService.class);
+            final SnapshotsInfoService snapshotsInfoService = getFollowerCluster().getCurrentMasterNodeInstance(SnapshotsInfoService.class);
 
-            final PlainActionFuture<IndexRoutingTable> waitForRestoreInProgress = PlainActionFuture.newFuture();
+            final PlainActionFuture<List<Long>> waitForAllShardSnapshotSizesFailures = PlainActionFuture.newFuture();
             final ClusterStateListener listener = event -> {
                 RestoreInProgress restoreInProgress = event.state().custom(RestoreInProgress.TYPE, RestoreInProgress.EMPTY);
                 if (restoreInProgress != null
                     && restoreInProgress.isEmpty() == false
                     && event.state().routingTable().hasIndex(followerIndex)) {
-                    waitForRestoreInProgress.onResponse(event.state().routingTable().index(followerIndex));
+                    try {
+                        final IndexRoutingTable indexRoutingTable = event.state().routingTable().index(followerIndex);
+                        assertBusy(() -> {
+                            List<Long> sizes = indexRoutingTable.shardsWithState(ShardRoutingState.UNASSIGNED).stream()
+                                .filter(shard -> shard.unassignedInfo().getLastAllocationStatus() == AllocationStatus.FETCHING_SHARD_DATA)
+                                .sorted(Comparator.comparingInt(ShardRouting::getId))
+                                .map(shard -> snapshotsInfoService.snapshotShardSizes().getShardSize(shard))
+                                .filter(Objects::nonNull)
+                                .collect(Collectors.toList());
+                            assertThat(sizes, hasSize(numberOfShards));
+                            waitForAllShardSnapshotSizesFailures.onResponse(sizes);
```

Comment:
Should this go outside the assertBusy()?

Reply:
Yes, actually there's no much value to capture the non-null sizes here and check again later that they are all equal to `UNAVAILABLE_EXPECTED_SHARD_SIZE`. We can check everything here, so I pushed 7c4480bcf0af405532382578f418ba634fa5da0f